packages/x-charts/src/ChartsLegend/ChartsLegend.tsx

```diff
@@ -207,3 +216,37 @@ export function ChartsLegend(inProps: ChartsLegendProps) {
     </ChartsLegendRoot>
   );
 }
+
+export function ChartsLegend(inProps: ChartsLegendProps) {
+  const props: DefaultizedChartsLegendProps = useThemeProps({
+    props: { ...defaultProps, ...inProps },
+    name: 'MuiChartsLegend',
+  });
+
+  const { position, direction, offset, hidden, slots, slotProps } = props;
+  const theme = useTheme();
+  const classes = useUtilityClasses({ ...props, theme });
+
+  const drawingArea = React.useContext(DrawingContext);
+  const series = React.useContext(SeriesContext);
+
+  if (hidden) {
+    return null;
+  }
+
+  const seriesToDisplay = getSeriesToDisplay(series);
+
+  const ChartLegendRender = slots?.legend ?? DefaultChartsLegend;
+  return (
```

Comment:
Is there something preventing us from using `useSlotProps`? 

This would allow supporting the callback format of the slot props and would apply some other smart behavior.

Reply:
There is one potential reason: the early return before `getSeriesToDisplay(series)` to avoid managing data if the legend is hidden which can not be done if we use a hook since it would make the number of hooks varry.

But the optimization does not worth it

Reply:
Same for the other slots on the PR :angel:

Reply:
Are you sure?

For the legend it could make sens since (with the tooltip) it's textual information that could get some props for customization

For line, area, scatter, they are into nested loops (loop other stacking group and loop other series).
So either:
- I use this hook at the top, but then the props mixing is kind of useless because at that moment we don't have access to the props specific to a given item
- I add an intermediate components to manage the slot/slotProps

The second solution seems more interesting but I don'gt thing slotsProps is the most interesting part. I believe users will mostly use `slots` for charts components, because those components are small enough to be rewritten for customization.

Reply:
> For line, area, scatter, they are into nested loops (loop other stacking group and loop other series).

You have a similar scenario [for the `day` slot](https://github.com/mui/mui-x/blob/2592e97a69fc85b2c10304b0cf77573cab653b2b/packages/x-date-pickers/src/DateCalendar/DayCalendar.tsx#L222)

My main issue here is consistency.
If most of our slots support the callback version but some do not, I can lead to frustration.

Reply:
May it could be moved into the `BarElement` (same for line and area). For Scatter and Marker since they are a lot, we could do it but by placing the slot/slotProps to at the item level but the series one.

Reply:
I'm fine with merging as is if you want and reconsider later :+1:

Reply:
I made an attempt here. If it's ok for you will duplicate it for other components when possible

https://github.com/mui/mui-x/pull/9744/commits/5dcf4acd4acc050019be01198d4ca7de3faa25b1

Reply:
This make a lot of sense to me